src/price_data.py

```diff
@@ -326,3 +328,133 @@ def get_cost(self, tr: Union[Operation, SoldCoin], reference_coin: str = config.
+        if self.exchange.has['fetchOHLCV']:
+            sleep(self.exchange.rateLimit / 1000)  # time.sleep wants seconds
+            # get 2min before and after range
+            return self.exchange.fetch_ohlcv(symbol, '1m', start-1000*60*2, max(int((stop-start)/1000/60)+5, 1))
+        else:
+            logging.warning(
+                "fetchOHLCV not implemented on exchange, skipping priceloading using ohlcv")
+            raise None
+
+    def initialize_ccxt(self):
+        exchange_id = 'binance'
+        exchange_class = getattr(ccxt, exchange_id)
+        self.exchange = exchange_class()
+        self.markets = []
+        markets = self.exchange.fetch_markets()
+
+        for market in markets:
+            # may not apply for all exchanges, currently works for binance
+            # caches a list of all pairs on the exchange
+            self.markets.append(market["symbol"].split("/"))
+
+    def _get_binance_bulk_pair_list(self, reference_coin: str = config.FIAT, coin) -> list:
+
+        def get_pair(coin, reference_coin:str):
+
+            for market in self.markets:
+                if market[0] == coin and market[1] == reference_coin:
+                    return [coin, reference_coin, False]  # False=not inverted
+                elif market[1] == coin and market[0] == reference_coin:
+                    return [reference_coin, coin, True]  # True=inverted
+
+        pair = get_pair(coin, reference_coin)
+        if not pair:
+
+            for market in self.markets:
+                pair = get_pair(market[1], reference_coin)
+
+                if pair:
+                    if market[0] == coin:
+                        return [[market[0], market[1], False], pair]
+                    elif market[1] == coin:
+                        return [[market[1], market[0], True], pair]
+        else:
+            return [pair, ]
+
+    def _get_binance_bulk_pair_data(self, operations: list, symbol: str, invert: str=False) ->list:
+        timestamps = []
+        timestamppairs = []
+        data = []
+
+        for op in operations:
+            timestamps.append(op.utc_time)
+
+        while len(timestamps) > 0:
+            timestamp = timestamps.pop(0)
+
+            if len(timestamppairs) > 0 and timestamppairs[-1][0]+datetime.timedelta(minutes=995) > timestamp:
+                timestamppairs[-1].append(timestamp)
+            else:
+                timestamppairs.append([timestamp, ])
+
+        for batch in timestamppairs:
+            # ccxt works with timestamps in milliseconds
+            last = int(max(batch).timestamp() * 1000)
+            first = int(min(batch).timestamp() * 1000)
+
+            if invert:
+                tempdata = list(
+                    map(lambda x: (x[0], 1/((x[1]+x[4])/2)), self.get_candles(first, last, symbol)))
+            else:
+                tempdata = list(
+                    map(lambda x: (x[0], (x[1]+x[4])/2), self.get_candles(first, last, symbol)))
```

Comment:
ccxt returns a tuple containing (timestamp in ms,Open,High,Low,Close,Volume) and we want to compute the avg price from Open/Close (maybe High/Low im currently not sure which one is better, because high/low can be a little misleading in some situations):
![image](https://user-images.githubusercontent.com/34819304/110201642-759d3b00-7e64-11eb-9d7d-c21c200885c7.png)
(ADA/BTC from 8:30 UTC)

From a python standpoint return None is actually not needed because with no return value the return is always None

Reply:
Yeah, lets stay with open/close for now, instead of high/low. I was not aware of how the API works.

---

"Explicit is better than implicit."
But you are right, the `return None` is not required there.